# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


## Carga de datos y Análisis descriptivo

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [2]:
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_1.head(10)

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
5  HHckp  -3.327590  -2.205276  3.003647   84.038886
6  h5Ujo -11.142655 -10.133399  4.002382  110.992147
7  muH9x   4.234715  -0.001354  2.004588   53.906522
8  YiRkx  13.355129  -0.332068  4.998647  134.766305
9  jG6Gi   1.069227 -11.025667  4.997844  137.945408

In [3]:
data_2 = pd.read_csv('/datasets/geo_data_2.csv')
data_2.head(10)

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
5  LzZXx -0.758092  0.710691  2.585887   90.222465
6  WBHRv -0.574891  0.317727  1.773745   45.641478
7  XO8fn -1.906649 -2.458350 -0.177097   72.480640
8  ybmQ5  1.776292 -0.279356  3.004156  106.616832
9  OilcN -1.214452 -0.439314  5.922514   52.954532

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Excelente inicio. La carga de los datasets se realiza correctamente y se visualizan las primeras filas, lo que permite verificar la estructura de los datos y comprender su composición desde el principio. Una base sólida para el análisis posterior.  
</div>


## geo_data_0

In [4]:
# se cargan los datos a la variable data_0
data_0 = pd.read_csv('/datasets/geo_data_0.csv')
data_0.head(10)

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
5  wX4Hy  0.969570  0.489775 -0.735383   64.741541
6  tL6pL  0.645075  0.530656  1.780266   49.055285
7  BYPU6 -0.400648  0.808337 -5.624670   72.943292
8  j9Oui  0.643105 -0.551583  2.372141  113.356160
9  OLuZU  2.173381  0.563698  9.441852  127.910945

In [5]:
# 2.1
X = data_0.drop(columns=['id','product'])
y = data_0['product']

In [6]:
# 2.1
X_train, X_val, y_train, y_val = train_test_split(X,y,test_size=.25,random_state=12345)

In [7]:
# 2.2
model = LinearRegression()
model.fit(X_train,y_train)

LinearRegression()

In [8]:
# 2.3
y_pred_data0 = model.predict(X_val)
y_pred_data0

array([ 95.89495185,  77.57258261,  77.89263965, ...,  61.50983303,
       118.18039721, 118.16939229])

In [9]:
# 2.4
rmse_data0 = mean_squared_error(y_val,y_pred_data0,squared=False)
rmse_data0

37.5794217150813

In [10]:
# 2.4
mean_pred_data0 = y_pred_data0.mean()
mean_pred_data0

92.59256778438035

In [11]:
# 2.4
mean_real_data0 = y_val.mean()
mean_real_data0

92.07859674082927

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Muy buena implementación del modelo lineal y de la separación de los datos en entrenamiento y validación. Los cálculos de métricas y promedios muestran un control adecuado del rendimiento del modelo y una comprensión clara del flujo de trabajo en aprendizaje supervisado.  
</div>


## 2.5
EL modelo hace predicciones muy cercanas a los datos de validación y la raíz del error cuadratico medio nos dice que hay una variabilidad de 37.6 mil barriles.

In [12]:
# 2.6


def entreno_validacion(df):
    X_f = df.drop(columns=['id','product'])
    y_f = df['product']
    X_trainf,X_valf,y_trainf,y_valf = train_test_split(
        X_f,y_f,test_size=.25,random_state=12345
    )
    modelf = LinearRegression()
    modelf.fit(X_trainf,y_trainf)
    
    y_pred_f = modelf.predict(X_valf)
    rmse_f = mean_squared_error(y_valf,y_pred_f,squared=False)
    mean_pred_f = y_pred_f.mean()
    mean_real_f = y_valf.mean()

    results = {
        "rmse":rmse_f,
        "mean_pred":mean_pred_f,
        "mean_real":mean_real_f,
        "y_val":y_valf.reset_index(drop=True),
        "y_pred":pd.Series(y_pred_f),
        "model":model
    }
    return results

In [21]:
# 2.6
resultado_0 = entreno_validacion(data_0)
resultado_0

{'rmse': 37.5794217150813,
 'mean_pred': 92.59256778438035,
 'mean_real': 92.07859674082927,
 'y_val': 0         10.038645
 1        114.551489
 2        132.603635
 3        169.072125
 4        122.325180
             ...    
 24995    170.116726
 24996     93.632175
 24997    127.352259
 24998     99.782700
 24999    177.821022
 Name: product, Length: 25000, dtype: float64,
 'y_pred': 0         95.894952
 1         77.572583
 2         77.892640
 3         90.175134
 4         70.510088
             ...    
 24995    103.037104
 24996     85.403255
 24997     61.509833
 24998    118.180397
 24999    118.169392
 Length: 25000, dtype: float64,
 'model': LinearRegression()}

In [22]:
# 2.6
resultado_1 = entreno_validacion(data_1)
resultado_1

{'rmse': 0.893099286775617,
 'mean_pred': 68.728546895446,
 'mean_real': 68.72313602435997,
 'y_val': 0         80.859783
 1         53.906522
 2         30.132364
 3         53.906522
 4          0.000000
             ...    
 24995    137.945408
 24996    110.992147
 24997    137.945408
 24998     84.038886
 24999     53.906522
 Name: product, Length: 25000, dtype: float64,
 'y_pred': 0         82.663314
 1         54.431786
 2         29.748760
 3         53.552133
 4          1.243856
             ...    
 24995    136.869211
 24996    110.693465
 24997    137.879341
 24998     83.761966
 24999     53.958466
 Length: 25000, dtype: float64,
 'model': LinearRegression()}

In [23]:
# 2.6
resultado_2 = entreno_validacion(data_2)
resultado_2

{'rmse': 40.02970873393434,
 'mean_pred': 94.96504596800489,
 'mean_real': 94.88423280885438,
 'y_val': 0         61.212375
 1         41.850118
 2         57.776581
 3        100.053761
 4        109.897122
             ...    
 24995     28.492402
 24996     21.431303
 24997    125.487229
 24998     99.422903
 24999    127.445075
 Name: product, Length: 25000, dtype: float64,
 'y_pred': 0         93.599633
 1         75.105159
 2         90.066809
 3        105.162375
 4        115.303310
             ...    
 24995     78.765887
 24996     95.603394
 24997     99.407281
 24998     77.779912
 24999    129.032417
 Length: 25000, dtype: float64,
 'model': LinearRegression()}

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Muy bien estructurada la función de entrenamiento y validación. La automatización del proceso y la obtención de métricas para cada dataset muestran un dominio creciente de la modularización y la evaluación de modelos en regresión.  
</div>


------

## 3.-

In [24]:
# 3.1
region_results = {
    "region_0":resultado_0,
    "region_1":resultado_1,
    "region_2":resultado_2
}

In [25]:
usd_per_unit = 4_500
top_wells = 200
budget_total = 100_000_000
cost_per_well = budget_total / top_wells
breakeven_units_per_well = cost_per_well / usd_per_unit

In [26]:
def summarize_region(res:dict,name:str):
    mean_pred = res['mean_pred']
    mean_real = res['mean_real']
    rmse = res['rmse']
    return {
        "region":name,
        "rmse":rmse,
        "mean_pred":mean_pred,
        "mean_real":mean_real,
        "breakeven_units_per_well":breakeven_units_per_well,
        "pred_minus_breakeven": mean_pred - breakeven_units_per_well,
        "real_minus_breakeven": mean_real - breakeven_units_per_well,
        "meets_breakeven_pred": mean_pred >= breakeven_units_per_well,
        "meets_breakeven_real": mean_real >= breakeven_units_per_well
    }

In [28]:
summary_df = pd.DataFrame(
    [summarize_region(res,name) for name,res in region_results.items()]
).set_index("region")
summary_df

rmse  mean_pred  mean_real  breakeven_units_per_well  \
region                                                                
region_0  37.579422  92.592568  92.078597                111.111111   
region_1   0.893099  68.728547  68.723136                111.111111   
region_2  40.029709  94.965046  94.884233                111.111111   

          pred_minus_breakeven  real_minus_breakeven  meets_breakeven_pred  \
region                                                                       
region_0            -18.518543            -19.032514                 False   
region_1            -42.382564            -42.387975                 False   
region_2            -16.146065            -16.226878                 False   

          meets_breakeven_real  
region                          
region_0                 False  
region_1                 False  
region_2                 False

Ninguna región supera el umbral de equilibrio promedio ni por los datos predichos ni por los datos reales pero no se descartan las regiones todavía ya que se van a evaluar los 200 mejores pozos.

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Excelente integración de los resultados regionales y su comparación con el punto de equilibrio. El análisis es claro y lógico, mostrando comprensión tanto del aspecto técnico como del criterio económico del proyecto.  
</div>


--------

## 4.-

In [30]:
def top200_profit(res):
    dfv = pd.DataFrame({"y_val":res["y_val"],"y_pred":res["y_pred"]})
    top200 = dfv.nlargest(top_wells,"y_pred")
    revenue = usd_per_unit * top200["y_val"].sum()
    profit = revenue - budget_total
    summary = {
        "count_top": len(top200),
        "sum_y_val_top200":top200["y_val"].sum(),
        "sum_y_pred_top200":top200["y_pred"].sum(),
        "mean_y_val_top200":top200["y_val"].mean(),
        "mean_y_pred_top200":top200["y_pred"].mean(),
        "profit_usd":profit
    }
    return top200,summary

In [31]:
results = {name:top200_profit(res) for name, res in region_results.items()}

In [33]:
summary_df = pd.DataFrame({name:r[1] for name,r in results.items()}).T.sort_values("profit_usd",ascending=False)
summary_df

count_top  sum_y_val_top200  sum_y_pred_top200  mean_y_val_top200  \
region_0      200.0      29601.835651       31102.330839         148.009178   
region_2      200.0      28245.222141       29603.898658         141.226111   
region_1      200.0      27589.081548       27746.026782         137.945408   

          mean_y_pred_top200    profit_usd  
region_0          155.511654  3.320826e+07  
region_2          148.019493  2.710350e+07  
region_1          138.730134  2.415087e+07

La mejor región es la 0 ya que es la que entrega la mayor ganacia potencial ($33.21 M) y el mayor volumen promedio por pozo.

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Excelente trabajo al identificar la región más rentable mediante el análisis de los 200 mejores pozos. El enfoque es preciso, con una buena combinación de criterios técnicos y económicos que fortalecen la interpretación de resultados.  
</div>


-------

## 5.-

In [34]:
n_boot = 1000

In [35]:
def bootstrap(res):
    df = pd.DataFrame({"y_val": res["y_val"], "y_pred": res["y_pred"]})
    profits = []
    for _ in range(n_boot):
        sample = df.sample(500, replace=True)
        top = sample.nlargest(top_wells, "y_pred")
        revenue = usd_per_unit * top["y_val"].sum()
        profit = revenue - budget_total
        profits.append(profit)
    profits = np.array(profits)
    return {
        "mean_profit": profits.mean(),
        "ci_low": np.percentile(profits, 2.5),
        "ci_high": np.percentile(profits, 97.5),
        "risk_loss": (profits < 0).mean() * 100
    }

In [36]:
results = {
    "region_0": bootstrap(resultado_0),
    "region_1": bootstrap(resultado_1),
    "region_2": bootstrap(resultado_2)
}

In [37]:
for r, v in results.items():
    print(f"{r}:")
    print(f"  Ganancia media: ${v['mean_profit']:,.0f}")
    print(f"  IC95%: [${v['ci_low']:,.0f}, ${v['ci_high']:,.0f}]")
    print(f"  Riesgo de pérdida: {v['risk_loss']:.2f}%\n")

region_0:
  Ganancia media: $3,908,065
  IC95%: [$-1,187,709, $9,002,204]
  Riesgo de pérdida: 6.80%

region_1:
  Ganancia media: $4,571,469
  IC95%: [$459,915, $8,615,557]
  Riesgo de pérdida: 1.40%

region_2:
  Ganancia media: $3,937,975
  IC95%: [$-1,720,922, $9,363,611]
  Riesgo de pérdida: 9.00%



La región 1 es la única que cuenta con un riesgo aceptable y ofrece la mayor ganancia media entre las que cumplen el umbral.
No coincide con el punto 4.3 ya que la variabilidad y los mejores promedios los tiene la región 1 y no la 0.

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Muy buena aplicación del método bootstrap para evaluar la incertidumbre y el riesgo de pérdida. La interpretación de los intervalos de confianza y del riesgo demuestra un entendimiento sólido de la estadística aplicada al análisis de rentabilidad.  
</div>


# Comentario General del Revisor

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a> 
    
¡Felicidades! Tu proyecto está **aprobado**. Has demostrado un excelente manejo del flujo completo de análisis de datos, desde la carga y exploración inicial hasta la modelación, evaluación y análisis de riesgo.  
A continuación, destaco los puntos positivos:  

#### Puntos Positivos:

* **Procesamiento de datos:** Implementaste correctamente la carga, limpieza y preparación de los datasets, garantizando consistencia y orden lógico en el proceso.
* **Modelado y validación:** Aplicaste modelos lineales con métricas claras y bien interpretadas, mostrando dominio de las herramientas de *scikit-learn*.
* **Automatización de funciones:** El uso de funciones reutilizables para entrenar, evaluar y comparar regiones refleja pensamiento analítico y buenas prácticas de programación.
* **Análisis económico y estadístico:** Incorporaste correctamente cálculos de rentabilidad y simulaciones bootstrap para evaluar riesgo, integrando criterios técnicos y financieros de forma precisa.

Tu trabajo evidencia rigor técnico y comprensión del contexto aplicado. Has realizado un análisis completo, coherente y bien documentado. Continúa profundizando en optimización de modelos y en visualizaciones para comunicar resultados con mayor impacto.

</div>
